Enable Subscription to invoke callbacks sequentially and synchronously

AvenDonn opened the pull request:
Pertaining to the discussion started in Issue #957, this Pull Request seeks to add capabilities for Subscription to sequentially invoke its MonitoredItem and FastDataChange callbacks to maintain order properly. The key goal is to avoid out-of-order monitored items. To this effect, the Sequential Publishing capability is added to Subscription
Resolves #957 

### SequentialPublishing
A toggle which forces Subscription to raise callbacks only in a sequential order by the SequenceNumber of the incoming message.
It limits the number of tasks that can process publish responses to 1, and enforces a strict +1 Sequence Number requirement before releasing a message.
### Technical details

I have written some notes below on my technical considerations. I recommend examining the code before reading this section, as this section serves to answer some of the questions you may have while reviewing the code. 

#### Backwards compatibility
The new properties have been added as DataMembers at the end to maintain backwards compatibility with the DataContract, where such may be used. The relevant copy constructor has also been updated.

Both features are disabled by default, so the existing behavior is the default. Users must explicitly set these properties on the Subscription object to enable these new features. 

#### Leveraging existing "time machine" in SaveMessageInCache
There is already a mechanism for re-ordering messages that entered in the wrong order. This change makes use of that existing "time machine" which sorts messages into chronological order (by the sequence number) to only pull off messages with a proper +1 sequence number each time.

#### KeepAlive messages advancing sequence number
KeepAlive messages do not contain notifications, and do not enter the proper branch in the code to be pulled out. They will not interrupt sequential flow. Since the next message in the sequence with data will "re-use" that sequence number, it can be expected to maintain sequence. 

#### Delayed messages
When sequential publishing is enabled, if a message is genuinely missing from the sequence, it will "hold up" the messages until it either arrives or is pulled out of the incoming messages list by being too old. In either case, it is considered "processed" for the sequence purposes at that point and the rest of the messages may proceed. 

The automatic republish request mechanism is also leveraged here for that purpose as well.

#### SequenceNumber roll-over after max value
As specified in the OPC UA spec, at 1ms publishing, this would take almost 50 days of uptime. Naturally at any "reasonable" publishing interval this time is even longer. The matter itself is also not addressed throughout the whole existing code that handled it before this change. (Though some consideration can be seen in some places, such as placing messages at the end of the incoming message linked list if no better place was found) 

#### Locking on m_cache to get the semaphore
I've considered making a separate locking object for the semaphore management, but decided it is not needed. The message worker would need to obtain m_cache lock later anyway, and momentarily obtaining it is not harmful for the overall flow.
Only in cases of serious contention between new PublishResponses coming in and workers attempting to work would this cause any problems.

#### Callbacks that "take a long time"
This change will naturally suffer from callbacks that take a long time to invoke. In our use case, each callback message is passed into an ActionBlock for handling, so our callbacks are fast. But Other clients may invoke much more code on these callbacks.
Sequential publishing means this capacity is possibly bottlenecked. Sequential publishing should be used for its intended purpose and with callbacks that do not hold up their calling thread for a long time. To ensure proper sequential callbacks, they cannot just be started in order, they have to be processed fully in order. 

#### Changing this property while the Subscription is active
There is limited support to on-the-fly changes to this property, and this is not the intended use case. It is meant to be set once and ideally never changed while messages are being processed.

If users need certain items properly sequenced and some items they can accept out-of-order or ignore outdated messages, they can do this by defining two separate Subscription objects, same as they would if they needed different Publishing Intervals. 

#### Why semaphore?
At first I had considered using ActionBlock from TPL DataFlow and limiting its concurrency to the number of desired message workers, but I had opted not to add a dependency on DataFlow only for this purpose. (Referencing all of DataFlow just for ActionBlock's automatic concurrency limiting is wasteful.

I had also considered using a Limited Concurrency Task Scheduler implementation and queueing the message workers on that instead of the normal ThreadPool via Task.Run, but that is similarly more complex than it needs to be.

#### Why async Task? What was wrong with Void?
While being on a new Task from the ThreadPool anyway, the worker would still occupy a task from the ThreadPool if it was using the synchronous Wait() method, which is a blocking call. By using WaitAsync, the ThreadPool task goes "back" to the ThreadPool until the semaphore is released. 

### Disclaimer - I am not an OPC UA expert
Please do not assume I have "done all my homework" with regard to these changes. I have attempted to learn as much as I can in the time I was working on this change, but I fully expect to have missed a few spots and specifics of the OPC UA protocol.

I have done some testing in our application to verify the out-of-order problem is solved, and that no significant delay is introduced by the additional work done to facilitate this.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/OPCFoundation/UA-.NETStandard?pullRequest=1493) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/OPCFoundation/UA-.NETStandard?pullRequest=1493) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/OPCFoundation/UA-.NETStandard?pullRequest=1493) it.</sub>

mregen commented on Libraries/Opc.Ua.Client/Subscription.cs:
```diff
@@ -177,10 +183,13 @@ public void Dispose()
         /// An overrideable version of the Dispose.
         /// </summary>
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2213:DisposableFieldsShouldBeDisposed", MessageId = "m_publishTimer")]
```
please fix nameof here too

mregen commented on Libraries/Opc.Ua.Client/Subscription.cs:
```diff
@@ -705,6 +778,8 @@ public void Create()
 
             AdjustCounts(ref revisedKeepAliveCount, ref revisedLifetimeCounter);
 
+            ManageMessageWorkerSemaphore();
```
the new createasync may need this call too

AvenDonn replied:
Ironically this call is functionally useless, it's in place in case the defaults ever need behavior to manage the worker semaphore. 

By default, workers are unlimited and sequential publishing is off. This means this method will leave the semaphore as null, as it started.

However, for avoiding accidental future issues (if for whatever reason, the default number of workers will be some number or sequential publishing will be the default behavior, which I don't think I recommend) then this will "protect" from that.

I'll add it to the new CreateAsync as well. Thanks for merging your recent Async library changes by the way, looked like a non-trivial merge.

mregen commented on Libraries/Opc.Ua.Client/Subscription.cs:
```diff
@@ -724,6 +799,29 @@ public void Create()
 
             ChangesCompleted();
         }
+        private void ManageMessageWorkerSemaphore()
```
I *think* the private methods are in a seperate region?

AvenDonn replied:
I'll have a look again

mregen commented on Libraries/Opc.Ua.Client/Subscription.cs:
```diff
@@ -724,6 +799,29 @@ public void Create()
+                    : m_maxMessageWorkers;
+
+                if (maxWorkers < 0)
+                    maxWorkers = 0; //Changing from one invalid value to another shouldn't make it update the semaphore
```
our coding style is Allman (always brackets)

mregen commented on Libraries/Opc.Ua.Client/Subscription.cs:
```diff
@@ -1240,17 +1338,22 @@ public IList<MonitoredItem> DeleteItems()
+                        if (entry.SequenceNumber == m_lastSequenceNumberProcessed + 1)
+                        {
+                            if (!entry.Processed)
+                                Utils.Trace("Subscription {0} skipping PublishResponse Sequence Number {1}", Id, entry.SequenceNumber);
```
add brackets

mregen commented on Libraries/Opc.Ua.Client/Subscription.cs:
```diff
@@ -1664,10 +1767,33 @@ private void AdjustCounts(ref uint keepAliveCount, ref uint lifetimeCount)
+            {
+                try
+                {
+                    await semaphore.WaitAsync();
```
I wonder if Tasks which are waiting here are really released in order, have you checked that itis guaranteed?

AvenDonn replied:
The order of the tasks being released is irrelevant. The worker tasks are fungible, doesn't matter which one is released. When sequential processing is enabled, only one will be active at a given time. 

The mechanism for ensuring order is how they pull items off the m_incomingMessages linked list, which gets "ordered" when new messages are inserted into it by SaveMessageInCache. 

Order of threads doesn't matter here.

mregen replied:
ok, got it, this is just the worker which pulls the next item...

mregen reviewed (commented):
Hi @AvenDonn , I see your approach here, it looks ok, but relies on the tasks to enter the semaphore  in the order they were blocked. However, [docs ](https://docs.microsoft.com/en-us/dotnet/standard/threading/semaphore-and-semaphoreslim)state: 

> There is no guaranteed order, such as first-in, first-out (FIFO) or last-in, first-out (LIFO), for blocked threads to enter the semaphore.

AvenDonn commented:
Hello @mregen, I have implemented your requested changes. 

However, I think there may be some missing coverage in testing this change for compliance with the protocol.
The tool testing the compliance should repeat the tests for monitored item values with sequential publishing enabled. I'm not sure how to make sure such coverage is present and would appreciate your help. 

However, I'm not sure it's very relevant to test arbitrary limits on max message workers.

mregen commented:
Hi @AvenDonn , currently our code coverage is just trying to execute as much code as possible, but is mostly not trying to execute special cases. I added a case to `AddSubscription` which you could decorate with specific tests for maxMessage workers and sequential publishing. 
Overall I think the code looks pretty good already and is definitely helping to improve the sequential ordering. If actual testing goes well we can merge it for the next release.

AvenDonn commented:
Hey @mregen, sounds good and glad you like the code. 
I've noted the unit test change and I'm not sure how to integrate an actual logical test for the actual value of the notification here, but given no previous coverage of it existed I'm not sure it's my place to try and bodge together some kind of coverage. The important part is setting these new properties and making sure nothing else is affected, which this does.

Glad to hear this is a candidate for the next release!

mregen commented:
Hi @AvenDonn , while discussing the fix a question came up. 
What is the rationale for specifying the max number of threads seperately in the config, when the SequentialMessaging flag already sets it to one?

Thanks,
Martin

mregen commented:
Hi @AvenDonn, thanks for your explanation... we had a discussion here and I think if we can reduce the test coverage for now its a good thing, so we can focus on testing the sequential publishing case only. Even better were if you had been able to provide a test case. At a certain point we would like to be in a state where we only accept contributions with a corresponding test. Way to go...
So could you pease remove the config for the number of workers?

mregen commented:
One more thing @AvenDonn : I was wondering what happens if a user sets `SequentialPublishing `and `DisableMonitoredItemCache`. Will the mechanism to recover with a republish still work?

AvenDonn commented:
@mregen I'm working on removing the max message workers feature now. I'm also going to see if I can write a test case. If I can mock a session to always provide messages out of order, I can verify sequential publishing fixes the order. 

As for your question, ```DisableMonitoredItemCache```  has no effect on the feature. The tracking is done independently of the callback being invoked on the ```MonitoredItem```. The intention is indeed that the ```FastDataChangeCallback``` and ```FastEventCallback``` are also affected.

Tracking of sequence happens in ```OnMessageReceived``` as messages are pulled out of the ```m_incomingMessages``` linked-list (Into which they were inserted in sequence by ```SaveMessageInCache```)
This tracking is done independently of whether callbacks are invoked. If the message was pulled out of the list, it's considered processed. (As before, indicated by the ```ii.Value.Processed = true;``` assignment.)

When in sequential publishing mode, the condition to enter this branch is true only if the sequence number is under +1 of the last seen sequence number. ```(!m_sequentialPublishing || ii.Value.SequenceNumber <= m_lastSequenceNumberProcessed + 1)```

If sequential publishing was enabled mid-way, all messages lower than the always-tracked ```m_lastSequenceNumberProcessed``` will still be allowed to go through. All subsequent messages must obey strict +1 sequence. 
If sequential publishing is disabled mid-way, then messages are always allowed to be processed, and ```m_lastSequenceNumberProcessed``` is still updated whenever ```ii.Value.SequenceNumber``` goes higher.

This is done by this code:
````
//Keep the last sequence number processed going up
if (ii.Value.SequenceNumber > m_lastSequenceNumberProcessed)
{
    m_lastSequenceNumberProcessed = ii.Value.SequenceNumber;
}
````

In sequential publishing mode, both conditions can only be true if the sequence number is +1.
In non-sequential-publishing (default) mode this value is only increased, to ensure that when sequential publishing is enabled, the next sequence message will actually increment it.

AvenDonn commented:
Hello @mregen. I have removed the arbitrary message workers control feature and I have also attempted to include some kind of unit test.

I am not satisfied with what I came up with for the unit test, as it is not deterministic and simply relies on running long enough for out-of-sequence to happen purely due to callbacks taking a long time.
I haven't found a convenient way to mock Session to spew messages out-of-sequence and observe Subscription adjust them properly. Even if I did that, the existing mechanism still properly takes care of it.

This test case does somewhat cover the multi-threading issue causing scenario, but doesn't very well illustrate the benefit of the strict +1 sequence number enforcement.

I'm not sure what's the best course of action here. If you feel the test is too convoluted, please remove it. I've made it an easily revertible dedicated commit. 
I'd appreciate with some help designing a better unit test for this case.

mregen reviewed (approved):
I debugged through the new code and I think its allworking as expected with minimal risk for regression, I may update the tests slightly, as it seems to be of not much value when we cannot inject publishrequests with non consecutive sequence number pattern. Great work @AvenDonn !

mregen commented:
Actually the new test is quite handy to validate high load and fast data callbacks...

mregen commented:
:shipit:

mregen commented:
Hi @AvenDonn , there is still some flakyness with the test case, as if on macOS we seem to hit the trigger condition sometimes which means there is a non sequential callback.

e.g. here https://opcfoundation.visualstudio.com/opcua-netstandard/_build/results?buildId=2340&view=results

any idea?

AvenDonn commented:
> Hi @AvenDonn , there is still some flakyness with the test case, as if on macOS we seem to hit the trigger condition sometimes which means there is a non sequential callback.
> 
> e.g. here https://opcfoundation.visualstudio.com/opcua-netstandard/_build/results?buildId=2340&view=results
> 
> any idea?

Hey @mregen, one of your changes "broke" a metric I'm relying on, the number of outstanding message workers. It did that because it disabled publishing before measuring it. That's not related to why it's failing though.

I see you've also removed the Thread.Sleep I've added in the callback code. This was meant to slow down the callback enough such that several callback invocations are running simultaneously. That was what was supposed to cause it to fail. 

I see each subscription tracks its items independently, so this isn't some odd collision of client handles somehow. But I wouldn't put it past some kind of possible aliasing issue coming from the server.
Why did you choose to create multiple subscriptions? To increase the chances of the issue occurring? Did you note the issue occurs when running without sequential publishing in this mode? Because I haven't been able to reproduce it given the current test.

Could you please test more with the original test I submitted?

mregen commented:
Thanks @AvenDonn , I will check your original tests again. They were definitely running for too long and it was hard to tell if they 'catched' an issue, so I basically repurposed them to become sort of a load tests. Then the sleeps are counterproductive. 
Having more subscriptions the idea is too have more publish requests queued on the server side, which may eventually lead to race condition.

AvenDonn commented:
Thanks @mregen. The server is not really the problem and adding more subscriptions won't add to the chance of reproducing since the ordering is on a per subscription basis anyway.

The trouble we've encountered came from the client, and was most likely stemming from concurrent execution of callbacks and possibly weird context switching and lock handling.
